### PR TITLE
sdk: introduce trait based `StorageAPI` (wip)

### LINKED ISSUE #11243: Introduce trait based storage API
This needs to follow #11240

because the block type is arbitrary we need a way to read and write arbitrary block data

### PR DESCRIPTION
**Pull request by joshieDo**
closes https://github.com/paradigmxyz/reth/issues/11243

tldr: each chain needs to write/read their own blocks. Eg. Not all chains may want `Ommers` as part of a block. These components are spread across different tables and need to be assembled.

Creating the PR to initiate a discussion. I think the first question is how to pass this from `ProviderFactory<N>` to `DatabaseProvider`. Right now it takes `<TX, Spec>` as generics.  https://github.com/paradigmxyz/reth/pull/11872 Adds a new `StateCommittment` generic. If we add a storage one, it's too much. Thoughts?

some options:
* unify `Spec and Storage` under yet another `Node` trait ? What about SC?
* pass `N`/`NodeTypes` from `ProviderFactory` (proposed here with `DatabaseProvider2`)
* just add `Storage` generic
* methods that require assembling a block take `ChainStorageReader`

### CODE REVIEW

**Comment by jenpaff** [comment]
> closes #11243
> 
> tldr: each chain needs to write/read their own blocks. Eg. Not all chains may want `Ommers` as part of a block. These components are spread across different tables and need to be assembled.
> 
> Creating the PR to initiate a discussion. I think the first question is how to pass this from `ProviderFactory<N>` to `DatabaseProvider`. Right now it takes `<TX, Spec>` as generics. #11872 Adds a new `StateCommittment` generic. If we add a storage one, it's too much. Thoughts?
> 
> some options:
> 
> * unify `Spec and Storage` under yet another `Node` trait ? What about SC?
> * pass `N`/`NodeTypes` from `ProviderFactory` (proposed here with `DatabaseProvider2`)
> * just add `Storage` generic
> * methods that require assembling a block take `ChainStorageReader`

how confident are you with our existing tests to make sure we're not breaking anything with this? I know there's been quite a bit of work around it

**Comment by mattsse** [comment]
yeah this progress will flag all the things that we need to do first to get unblocked

File: crates/ethereum/node/src/node.rs [hunk 1]
```diff
@@ -126,6 +132,61 @@ where
     }
 }
 
+/// Ethereum storage that implements [`ChainStorageReader`].
+#[derive(Debug, Default)]
+pub struct EthStorage;
```
Comment (mattsse) [inline crates/ethereum/node/src/node.rs:137]: I assume what we want here instead is a generic variant because for both eth and op this will look the same.

File: crates/storage/storage-api/src/chain.rs [hunk 2]
```diff
@@ -0,0 +1,71 @@
...
+        provider: &P,
+        id: BlockHashOrNumber,
+    ) -> ProviderResult<Option<<Self::Primitives as NodePrimitives>::Block>>
+    where
+        P: DBProvider<Tx: DbTx>
+            + TransactionsProvider
+            + BlockReader
+            + WithdrawalsProvider
+            + HeaderProvider
+            + BlockNumReader;
```
Comment (klkvr) [inline crates/storage/storage-api/src/chain.rs:30]: wondering if we can avoid those bounds on trait level? I guess we can make the entire trait generic over the provider?

Comment (mattsse) [reply crates/storage/storage-api/src/chain.rs:30]: yeah these are problematic. 
I believe @joshieDo also discussed changing to the just to the actual dbprovider instance so we don't need to enforce these traits on the storage trait itself
